### PR TITLE
feat(rome_js_analyze): useExportType

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

This rule implements some part of [consistent-type-exports](https://typescript-eslint.io/rules/consistent-type-exports/).
In contrast to the ESLint rule:

- we use `export { type }` by default. ESLint provides an option for this and split an export between a value export and a `export type`.

  For example, given the following code:

  ```ts
  interface I {}
  class C {}
  export { I, C }
  ```

  ESlint proposes the following fix:

  ```ts
  interface I {}
  class C {}
  export { C }
  export type { I }
  ```

  While this rules proposes the following fix:

  ```ts
  interface I {}
  class C {}
  export { type I, C }
  ```

- The ESLint rule seems aware of which re-export name is a type. Rome has no information on re-exported names.

## Test Plan

Tests included.

### CODE REVIEW

**Comment by Conaclos** [comment]
For now the rule has a simple check to avoid adding an inline type to a single named export. In other words `export { MyType }` is turned into `export type { MyType }` instead of `export { type MyType }`.

However, this does not apply to other situations. For instance `export { MyType, type A }` is turned into `export { type MyType, type A }`.

I am currently not satisfied with this incosnistency.
I have two proposals to be more consistent:

(1) always use inline types and provide another rule `useGroupedExportType` that factorize an export that contains only inline type exports into `export type {}`.

(2) Always factorize inline type imports into `export type`.
    In this case, the rule could integrate `useGroupedExportType`

I am more inclined for (2).
If we take this path, we should certainly introduce `useImportType` and integrate [useGroupedTypeImport](https://docs.rome.tools/lint/rules/useGroupedTypeImport/) in `useImportType`.

**Comment by Conaclos** [comment]
> What's the reasoning of diverging from eslint? It would be great to provide this 
> information

In the rule docs? or just here?

The ESLint rule was created before that _TypeScript_ supports inline type exports.
They expose an option to use inline type exports.

I based my choice on the principle of making the less change as possible in the fix. Moreover, I think it is more idiomatic of keeping types and values from a same module grouped.

**Comment by ematipico** [comment]
> In the rule docs? or just here?

Both. If we diverge we should use `Credits:`, not source. Small but important difference.

**Comment by ematipico** [comment]
> The second difference comes from technical limitation: this could require Rome to provide multi-file analysis.

Could you provide more information? I am not sure I understand

**Review by ematipico — CHANGES_REQUESTED** [review]

**Comment by Conaclos** [comment]
> > The second difference comes from technical limitation: this could require Rome to provide multi-file analysis.
> 
> Could you provide more information? I am not sure I understand

Sorry I was wrong. There is no difference here.
I though so that ESlint handled the following case:

```ts
export { X } from "mod" // X is a type and should thus be epxorted with the `export type`
```

This check require multi-file analysis (we have to know what `X` is).

**Comment by Conaclos** [comment]
@ematipico
Have you some input to share about https://github.com/rome/tools/pull/4751#issuecomment-1662505197?

**Comment by ematipico** [comment]
To be honest, I think we should **not** implement this rule *yet*, mostly because we won't be able to catch cases like `export type * as mod from "mod"`, which would result in some inconsistencies. 

Instead, as you suggested before, I would expand the rule https://docs.rome.tools/lint/rules/useGroupedTypeImport/ to apply it also to the export types

**Comment by Conaclos** [comment]
> To be honest, I think we should **not** implement this rule _yet_, mostly because we won't be able to catch cases like `export  type * as mod from "mod"`, which would result in some inconsistencies.

I am not sure to understand what you refer to. `export type * as mod from "mod"` is currently handled. This imports only types.

> Instead, as you suggested before, I would expand the rule https://docs.rome.tools/lint/rules/useGroupedTypeImport/ to apply it also to the export types

In fact, I suggested writing a new rule named `useImportType` that includes also `useGroupedTypeImport`.
Do you mean creating a rule that factorize export and import inline types?

**Comment by ematipico** [comment]
> > To be honest, I think we should **not** implement this rule _yet_, mostly because we won't be able to catch cases like `export  type * as mod from "mod"`, which would result in some inconsistencies.
> 
> I am not sure to understand what you refer to. `export type * as mod from "mod"` is currently handled. This imports only types.
> 

Sorry, bad example. I meant that when using the syntax `export * from mod`, we aren't able to determine if inside `mod` we have only types. Or, maybe it's fine if don't try to analyse this use case. 


> > Instead, as you suggested before, I would expand the rule https://docs.rome.tools/lint/rules/useGroupedTypeImport/ to apply it also to the export types
> 
> In fact, I suggested writing a new rule named `useImportType` that includes also `useGroupedTypeImport`.
> Do you mean creating a rule that factorize export and import inline types? 

Sorry, this is confusing. 

As long as we have the **one** rule that does one job and it's consistent, I think we're good. I didn't understand why you wanted to change the name, but it's **fine**, as long as **we don't have two rules** that do the same thing (one for imports and one for exports).

**Comment by ematipico** [comment]
One more thought, why don't we implement this logic in import sorting?? I think it's more appropriate to have it there.

**Comment by Conaclos** [comment]
> As long as we have the one rule that does one job and it's consistent, I think we're good. I didn't understand why you wanted to 
> change the name, but it's fine, as long as we don't have two rules that do the same thing (one for imports and one for exports).

Let's me dive in the details. Hopefully this will help you to understand my point.
I think about four distinct rules:

- [`useGroupedTypeImport`](https://docs.rome.tools/lint/rules/usegroupedtypeimport/)

  Factorize inline type imports.
  For example:

  ```diff
  - import { type X, type Y } from ""
  + import type { X, Y } from ""

  - import { type Z } from ""
  + import type { Z } from ""

    // cannot be factorized
    import { type U, V } from ""
  ```

- `useGroupedExportType` (or `useGroupedTypeExport`)

  Factorize inline type exports
  For example:

  ```diff
  - export { type X, type Y }
  + export type { X, Y }

  - export { type Z }
  + export type { Z }

    // cannot be factorized
    export { type U, V }
  ```

- `useImportType`

  Every import that only appears in a type position MUST be imported with `import type` or an inline type import.
  For example:

  ```diff
  - import { X, type Y } from ""
  + import { type T, type U } from ""

    export function f (x: X): Y {
        return x.g()
    }

    // here Z is used both as a value and a type
    import { Z } from ""
    export const C: Z = Z.C
  ```

- `useExportType` - current PR

  Every export that is defined as a type-only value MUST be exported with `export type` or an inline type export.
  For example:

  ```
    interface X {}
    interface Y {}
  - export { X, type Y }
  + export { type X, type Y }

    // a value can be exported as a type or a value
    class C {}
    class D {}
    export { C, type D }
  ```

Because fixing `useImportType` and `useExportType` could lead to code that can respectively trigger `useGroupedTypeImport` and `useGroupedExportType`, I proposed merging `useImportType` with `useGroupedTypeImport`, and `useExportType` with `useGroupedExportType`.  

> One more thought, why don't we implement this logic in import sorting?? I think it's more appropriate to have it there.

I didn't think about that. I think it is a good idea.
Although it is a style-related rule, I didn't see why someone could oppose turning `import { type }` into `import type`.
However, I wonder: do you think `useGroupedExportType` could also be part of the import organizer? This could extend the organizer to export statements?

By the way, `useImportType` and `useExportType` are still useful to verify that all type-only entities are properly exported and properly imported.
In the case of exports, we cannot perform cross-platform checks. However, I think the rule is valuable even without this check. The rule could be improved in the future.

File: crates/rome_js_syntax/src/export_ext.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    pub fn with_type_token(self, type_token: Option<JsSyntaxToken>) -> Self {
+        match self {
+            AnyJsExportNamedSpecifier::JsExportNamedShorthandSpecifier(specifier) => {
+                specifier.with_type_token(type_token).into()
+            }
+            AnyJsExportNamedSpecifier::JsExportNamedSpecifier(specifier) => {
+                specifier.with_type_token(type_token).into()
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/rome_js_syntax/src/export_ext.rs:89]: Many mutating methods are already present in [`rome_js_syntax/src/generated/nodes_mut.rs`](https://github.com/rome/tools/blob/main/crates/rome_js_syntax/src/generated/nodes_mut.rs).

File: crates/rome_js_syntax/src/import_ext.rs [hunk 2]
```diff
@@ -45,3 +49,69 @@ impl JsModuleSource {
...
+        match self {
+            AnyJsNamedImportSpecifier::JsBogusNamedImportSpecifier(_) => None,
+            AnyJsNamedImportSpecifier::JsNamedImportSpecifier(specifier) => specifier.type_token(),
+            AnyJsNamedImportSpecifier::JsShorthandNamedImportSpecifier(specifier) => {
+                specifier.type_token()
+            }
+        }
+    }
+
+    pub fn local_name(&self) -> Option<AnyJsBinding> {
```
[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/rome_js_syntax/src/import_ext.rs:75]: Unfortunately the unique `SyntaxError` variant (`SyntaxError::MissingRequiredChild`) is not a fit for this code?

File: crates/rome_js_syntax/src/export_ext.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_syntax/src/export_ext.rs:89]: Wouldn't make sense to follow the same pattern? (create a file with _mut postfix)What do you think?

File: crates/rome_js_syntax/src/import_ext.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_syntax/src/import_ext.rs:75]: Yeah it does, it's the only error we throw. A token either required or optional

File: crates/rome_js_syntax/src/export_ext.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/rome_js_syntax/src/export_ext.rs:89]: In fact, I don't see the benefit of separating it. IMHO the `mut` keyword already create a separation.

**Comment by ematipico** [comment]
@Conaclos 

Thank you for explaining in detail what you had in mind.

First, **import sorting** is _**not**_ about formatting but about **analyzing**. JavaScript is a language that has side effects; hence the order of the import statements matters. Most of the time, this is not an issue, but we should be aware of this distinction as developers that make developers tools for other fellow developers.

Having said that, the rules you proposed around `import type` can be implemented in the `organize_imports.rs` file, where we implement the algorithm for sorting imports. Doing so would also pressure the users to the feature inside Rome and teach them that sorting import statements inside **JavaScript** is about _analyzing_ the code and not formatting.

I agree that we can merge rules around `export type`. I don't have strong opinions between `export type A` and `export type { A }`.

**Pull request closed without merging**